### PR TITLE
feat: earth engine layers aggregation and refactor (DHIS2-9530)

### PR DESCRIPTION
**Pull request by turban**
Partly fixes: https://jira.dhis2.org/browse/DHIS2-9530

Related maps-gl PR which I would advice to review first: https://github.com/dhis2/maps-gl/pull/234

This PR improves the Google Earth Engine support in DHIS2 Maps, allowing data to be aggregated to organisation units. 

There is one more PR that could be reviewed after this: https://github.com/dhis2/maps-app/pull/1477 (showing earth engine calculations in the data table).

if you want to test this branch: 
- Checkout `feat/DHIS2-9530` branch of both maps-gl
- In maps-gl: `yarn install` / `yarn build` / `cd build` / `yarn link`
- In maps-app: `yarn link @dhis2/maps-gl`

After this PR there is one additional population layer (with age/sex groups) while nighttime lights is removed (outdated): 

Screenshot 2021-02-18 at 13 03 57

Before: 

Screenshot 2021-02-18 at 13 05 08

After this PR the Earth Engine layer dialogs have a new data tab: 

Screenshot 2021-02-18 at 13 12 47

Screenshot 2021-02-18 at 13 15 23

Screenshot 2021-02-18 at 13 15 43

Screenshot 2021-02-18 at 13 17 15

Screenshot 2021-02-18 at 13 17 50

Screenshot 2021-02-18 at 13 18 18

The new data tab is adapted to each dataset: 
- The layer name is shown at the top of the dialog
- A short layer description is shown before a generic description of how the layer can be used
- An "Aggregation method" select allows the user to set how the data can be aggregated. More info below. 
- The "Population age groups" has an additional select for age/sex groups. More info below. 

Allowed aggregation methods: 

Screenshot 2021-02-18 at 13 26 32

We provide a sensible default for each layer. 

For landcover the methods are different: 

Screenshot 2021-02-18 at 13 28 03

The reason is that this dataset contains multiple variables (landcover types), and this allows the user to compare them. Here we only allow one selection due to display restrictions. 

Period select after this PR: 

Screenshot 2021-02-18 at 13 31 47

Screenshot 2021-02-18 at 13 32 09

These are set by the source (GEE). We preselect the most recent period. A loading indicator is shown while available periods are loading: 

Screenshot 2021-02-18 at 13 38 55

We provide the same organisation unit select as for other layers:

Screenshot 2021-02-18 at 13 39 49

The second org unit level is selected by default (we do the same for thematic layers). 

The style tab contains the same options as previous versions: 

Screenshot 2021-02-18 at 13 42 07

We have added a loading spinner for Earth Engine Layers, as they often take time to process:

Screenshot 2021-02-18 at 13 43 14

This maps-gl PR includes screenshots of how they layer look and operate: https://github.com/dhis2/maps-gl/pull/234

### CODE REVIEW

File: src/components/edit/earthEngine/StyleSelect.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            }
+        },
+        [params, setParams]
+    );
+
+    return (
+        <div className={styles.flexColumnFlow}>
+            <div className={styles.flexColumn}>
+                <p>
+                    {i18n.t('Unit')}: {unit}
```
Comment (edoardo) [inline src/components/edit/earthEngine/StyleSelect.js:51]: Use interpolation:
```suggestion
                    {i18n.t('Unit: {{ unit }}', {unit})}
```

File: src/components/edit/earthEngine/styles/PeriodSelect.module.css [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+.loading {
+    display: flex;
+    align-items: center;
+    font-style: italic;
+    padding-left: var(--spacers-dp16);
+}
+
+.loading div {
+    display: inline-block;
+    margin-right: 24px;
```
Comment (edoardo) [inline src/components/edit/earthEngine/styles/PeriodSelect.module.css:24]:
```suggestion
    margin-right: var(--spacers-dp24);
```

File: src/components/edit/styles/LayerDialog.module.css [hunk 3]
```diff
@@ -46,16 +48,37 @@
     font-weight: bold;
 }
 
+.paragraph {
+    padding: var(--spacers-dp8) 0 var(--spacers-dp4);
+    font-size: 13px;
+    line-height: 16px;
```
Comment (edoardo) [inline src/components/edit/styles/LayerDialog.module.css:54]:
```suggestion
    line-height: var(--spacers-dp16);
```

File: src/components/map/layers/earthEngine/styles/EarthEnginePopup.module.css [hunk 4]
```diff
@@ -0,0 +1,63 @@
+.title {
+    font-weight: bold;
+    font-size: 16px;
+    padding-bottom: 16px;
```
Comment (edoardo) [inline src/components/map/layers/earthEngine/styles/EarthEnginePopup.module.css:4]: Use `spacers` from `ui` wherever possible

File: src/components/map/layers/styles/LayerLoading.module.css [hunk 5]
```diff
@@ -0,0 +1,13 @@
+.layerLoading {
+    display: flex;
+    align-items: center;
+    padding-right: 16px;
```
Comment (edoardo) [inline src/components/map/layers/styles/LayerLoading.module.css:4]: Same here.

File: src/loaders/earthEngineLoader.js [hunk 6]
```diff
@@ -1,196 +1,29 @@
...
     let layerConfig = {};
     let dataset;
+    let features;
+
+    if (orgUnits && orgUnits.length) {
+        const d2 = await getD2();
+        const displayProperty = getDisplayProperty(d2).toUpperCase();
+        const orgUnitParams = orgUnits.map(item => item.id);
+
+        features = await d2.geoFeatures
```
Comment (edoardo) [inline src/loaders/earthEngineLoader.js:21]: We should probably move this to `@dhis2/analytics` similar to the `analytics` part.

File: src/loaders/earthEngineLoader.js [hunk 7]
```diff
@@ -222,9 +55,36 @@ const earthEngineLoader = async config => {
...
+        source,
+        sourceUrl,
+        band,
+        bands,
+    } = layer;
+
+    const period = getPeriodFromFilter(filter);
+
+    const groups =
+        band && Array.isArray(bands)
```
Comment (edoardo) [inline src/loaders/earthEngineLoader.js:72]: Should this check for `bands.length`?

File: src/util/earthEngine.js [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+    return f
+        ? {
+              id: f.id || f.arguments[1],
+              name: f.name,
+              year: f.year,
+          }
+        : null;
+};
+const setAuthToken = async ({ client_id, access_token, expires_in }) =>
```
Comment (edoardo) [inline src/util/earthEngine.js:30]: `async` not necessary?

File: src/util/earthEngine.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        .ImageCollection(id)
+        .distinct('system:time_start')
+        .sort('system:time_start', false);
+
+    const featureCollection = ee
+        .FeatureCollection(imageCollection)
+        .select(['system:time_start', 'system:time_end'], null, false);
+
+    // Return first and last date for daily periods
+    if (periodType === 'Daily') {
```
Comment (edoardo) [inline src/util/earthEngine.js:92]: Should these be constants imported from somewhere?

File: src/util/favorites.js [hunk 10]
```diff
@@ -120,22 +122,32 @@ const models2objects = config => {
...
+            band,
+            aggregationType,
             filter,
-            periodName,
-        });
+        };
+
+        // Removes undefined keys before stringify
+        config.config = JSON.stringify(
+            Object.keys(eeConfig).forEach(
```
Comment (edoardo) [inline src/util/favorites.js:138]: Does this work?
Isn't `forEach` returning `undefined`?

**Review by edoardo — APPROVED** [review]
I think some comments need to be addressed.

File: src/components/edit/earthEngine/StyleSelect.js (hunk 1 above)
Comment (turban) [reply src/components/edit/earthEngine/StyleSelect.js]: Fixed.

File: src/components/edit/styles/LayerDialog.module.css (hunk 3 above)
Comment (turban) [reply src/components/edit/styles/LayerDialog.module.css]: Fixed.

File: src/components/map/layers/earthEngine/styles/EarthEnginePopup.module.css (hunk 4 above)
Comment (turban) [reply src/components/map/layers/earthEngine/styles/EarthEnginePopup.module.css]: Fixed.

File: src/components/map/layers/styles/LayerLoading.module.css (hunk 5 above)
Comment (turban) [reply src/components/map/layers/styles/LayerLoading.module.css]: Fixed.

File: src/loaders/earthEngineLoader.js (hunk 6 above)
Comment (turban) [reply src/loaders/earthEngineLoader.js:21]: Agreed :-)

File: src/loaders/earthEngineLoader.js (hunk 7 above)
Comment (turban) [reply src/loaders/earthEngineLoader.js]: Added.

File: src/util/earthEngine.js (hunk 9 above)
Comment (turban) [reply src/util/earthEngine.js]: Kept for now, but I removed the code for Daily period as it's not in use yet.

File: src/util/favorites.js (hunk 10 above)
Comment (turban) [reply src/util/favorites.js:138]: forEach is returning the object keys which are not undefined.

File: src/util/earthEngine.js (hunk 8 above)
Comment (turban) [reply src/util/earthEngine.js]: Removed.

File: src/components/edit/earthEngine/styles/PeriodSelect.module.css (hunk 2 above)
Comment (turban) [reply src/components/edit/earthEngine/styles/PeriodSelect.module.css]: Fixed.

File: src/util/favorites.js (hunk 10 above)
Comment (edoardo) [reply src/util/favorites.js:138]: 🤔  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#return_value

Comment (turban) [reply src/util/favorites.js:138]: Fixed in #1477 as this branch was already merged.

**Pull request merged**